File: src/views/MapView.js
/**
 * MapView.js - Handles all map rendering with Leaflet
 * Responsible for DOM manipulation and map visualization
 */
class MapView {
  constructor(containerId) {
    this.container = document.getElementById(containerId);
    this.map = null;
    this.layers = new Map();
    this.markers = new Map();
    this.initMap();
  }

  /**
   * Initialize Leaflet map
   */
  initMap() {
    if (!this.container) {
      console.error('Map container not found');
      return;
    }

    // Initialize map centered on Norway
    this.map = L.map(this.container, {
      zoom: 5,
      center: [60.4518, 8.4689],
      attributionControl: true
    });

    // Add base tiles
    L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
      attribution: '© OpenStreetMap contributors',
      maxZoom: 19
    }).addTo(this.map);
  }

  /**
   * Add GeoJSON layer to map
   * @param {string} layerId - Unique layer identifier
   * @param {object} geojson - GeoJSON FeatureCollection
   * @param {object} options - Styling and popup options
   */
  addGeoJSONLayer(layerId, geojson, options = {}) {
    const defaultStyle = {
      color: options.color || '#3388ff',
      weight: options.weight || 2,
      opacity: options.opacity || 0.8,
      fillOpacity: options.fillOpacity || 0.5
    };

    const layer = L.geoJSON(geojson, {
      style: defaultStyle,
      onEachFeature: (feature, layer) => {
        const popup = this.createPopup(feature.properties);
        layer.bindPopup(popup);
        
        // Add hover effects
        layer.on('mouseover', function() {
          this.setStyle({
            weight: (options.weight || 2) + 2,
            opacity: 1
          });
        });
        layer.on('mouseout', function() {
          this.setStyle(defaultStyle);
        });
      }
    });

    layer.addTo(this.map);
    this.layers.set(layerId, { layer, geojson });
    return layer;
  }

  /**
   * Create popup HTML from feature properties
   * @param {object} properties - Feature attributes
   * @returns {string} HTML popup content
   */
  createPopup(properties) {
    if (!properties || Object.keys(properties).length === 0) {
      return '<div class="popup-content"><p>No data available</p></div>';
    }

    let html = '<div class="popup-content"><table>';
    Object.entries(properties).forEach(([key, value]) => {
      html += `<tr><td><strong>${key}</strong></td><td>${value}</td></tr>`;
    });
    html += '</table></div>';
    return html;
  }

  /**
   * Add marker to map
   * @param {string} markerId - Unique marker identifier
   * @param {array} latlng - [latitude, longitude]
   * @param {object} options - Marker options (title, icon, etc.)
   */
  addMarker(markerId, latlng, options = {}) {
    const marker = L.marker(latlng, {
      title: options.title || 'Marker'
    }).bindPopup(options.popup || options.title);

    if (options.icon) {
      marker.setIcon(options.icon);
    }

    marker.addTo(this.map);
    this.markers.set(markerId, marker);
    return marker;
  }

  /**
   * Toggle layer visibility
   * @param {string} layerId - Layer identifier
   * @param {boolean} visible - Show/hide layer
   */
  toggleLayerVisibility(layerId, visible) {
    if (!this.layers.has(layerId)) return;

    const { layer } = this.layers.get(layerId);
    if (visible) {
      this.map.addLayer(layer);
    } else {
      this.map.removeLayer(layer);
    }
  }

  /**
   * Remove layer from map
   * @param {string} layerId - Layer identifier
   */
  removeLayer(layerId) {
    if (this.layers.has(layerId)) {
      const { layer } = this.layers.get(layerId);
      this.map.removeLayer(layer);
      this.layers.delete(layerId);
    }
  }

  /**
   * Update layer with new GeoJSON data
   * @param {string} layerId - Layer identifier
   * @param {object} geojson - New GeoJSON data
   */
  updateLayer(layerId, geojson) {
    if (this.layers.has(layerId)) {
      this.removeLayer(layerId);
      this.addGeoJSONLayer(layerId, geojson);
    }
  }

  /**
   * Fit map bounds to layer
   * @param {string} layerId - Layer identifier
   */
  fitToBounds(layerId) {
    if (this.layers.has(layerId)) {
      const { layer } = this.layers.get(layerId);
      this.map.fitBounds(layer.getBounds());
    }
  }

  /**
   * Get current map viewport
   * @returns {object} {center: [lat, lng], zoom: number}
   */
  getViewport() {
    const center = this.map.getCenter();
    return {
      center: [center.lat, center.lng],
      zoom: this.map.getZoom()
    };
  }

  /**
   * Set map viewport
   * @param {array} center - [latitude, longitude]
   * @param {number} zoom - Zoom level
   */
  setViewport(center, zoom) {
    this.map.setView(center, zoom);
  }

  /**
   * Draw search radius circle
   * @param {array} center - [latitude, longitude]
   * @param {number} radiusKm - Radius in kilometers
   */
  drawSearchRadius(center, radiusKm) {
    if (this.searchCircle) {
      this.map.removeLayer(this.searchCircle);
    }

    this.searchCircle = L.circle(center, {
      radius: radiusKm * 1000, // Convert to meters
      color: '#ff6b6b',
      fill: false,
      weight: 2,
      dashArray: '5, 5',
      opacity: 0.8
    }).addTo(this.map);
  }

  /**
   * Clear search visualization
   */
  clearSearchVisualization() {
    if (this.searchCircle) {
      this.map.removeLayer(this.searchCircle);
      this.searchCircle = null;
    }
  }
}

export default MapView;

File: app/views/map_view.py
"""
MapView.py - Handles map rendering with Folium
Converts geographic data into interactive web map visualizations
"""
import folium
from folium import plugins
from typing import Dict, List, Tuple, Optional


class MapView:
    def __init__(self, center: List[float], zoom_level: int):
        """
        Initialize Folium map
        :param center: [latitude, longitude]
        :param zoom_level: Initial zoom level
        """
        self.map = folium.Map(
            location=center,
            zoom_start=zoom_level,
            tiles='OpenStreetMap'
        )
        self.layers_group = {}

    def add_geojson_layer(self, layer_id: str, features: List[Dict], 
                         color: str = '#3388ff', popup_fields: List[str] = None):
        """
        Add GeoJSON layer to map with data-driven styling
        :param layer_id: Unique layer identifier
        :param features: List of GeoJSON features
        :param color: Default layer color (used if no specific style applies)
        :param popup_fields: Fields to show in popup
        """
        feature_group = folium.FeatureGroup(name=layer_id)
        
        for feature in features:
            geometry = feature.get('geometry', {})
            properties = feature.get('properties', {})
            
            # Data-driven styling based on property 'type'
            feature_color = color
            if 'type' in properties:
                type_val = properties['type']
                if type_val == 'Government Building':
                    feature_color = '#795548'  # Brown
                elif type_val == 'Port':
                    feature_color = '#2196f3'  # Blue
                elif type_val == 'Religious Building':
                    feature_color = '#9c27b0'  # Purple
                elif type_val == 'Infrastructure':
                    feature_color = '#607d8b'  # Grey
            
            if geometry['type'] == 'Point':
                coords = geometry['coordinates']
                
                # Create popup HTML
                popup_html = self._create_popup(properties, popup_fields)
                
                folium.CircleMarker(
                    location=[coords[1], coords[0]],  # (lat, lng)
                    radius=8,
                    popup=folium.Popup(popup_html, max_width=300),
                    color=feature_color,
                    fill=True,
                    fillColor=feature_color,
                    fillOpacity=0.7,
                    weight=2,
                    opacity=0.9
                ).add_to(feature_group)
                
            elif geometry['type'] == 'LineString':
                coords = [[c[1], c[0]] for c in geometry['coordinates']]
                folium.PolyLine(
                    coords,
                    color=feature_color,
                    weight=4,
                    opacity=0.8,
                    popup=folium.Popup(self._create_popup(properties, popup_fields), 
                                      max_width=300)
                ).add_to(feature_group)
        
        feature_group.add_to(self.map)
        self.layers_group[layer_id] = feature_group

    def draw_search_radius(self, center: Tuple[float, float], radius_km: float):
        """
        Draw search radius circle
        :param center: (latitude, longitude)
        :param radius_km: Radius in kilometers
        """
        folium.Circle(
            location=center,
            radius=radius_km * 1000,  # Convert to meters
            color='#ff6b6b',
            fill=False,
            weight=2,
            opacity=0.8,
            dash_array='5, 5'
        ).add_to(self.map)

    def add_layer_control(self):
        """Add layer control to map"""
        folium.LayerControl().add_to(self.map)

    def _create_popup(self, properties: Dict, fields: List[str] = None) -> str:
        """
        Create HTML popup content from feature properties
        :param properties: Feature attributes
        :param fields: Specific fields to show (None = all)
        :return: HTML string
        """
        if not properties:
            return "<div><p>No data available</p></div>"
        
        if fields:
            properties = {k: v for k, v in properties.items() if k in fields}
        
        html = "<div style='font-size: 12px; width: 200px;'><table style='width:100%;'>"
        for key, value in properties.items():
            html += f"<tr><td style='font-weight:bold;'>{key}:</td><td>{value}</td></tr>"
        html += "</table></div>"
        return html

    def get_html(self) -> str:
        """Get map as HTML string"""
        return self.map._repr_html_()

    def save(self, filepath: str):
        """Save map to HTML file"""
        self.map.save(filepath)
